### PR TITLE
Now, `reward_fn` lives in LLMRayActor.

### PR DESCRIPTION
**Pull request by finbarrtimbers**
This has a few advantages: 

1. Timing wise, we won't block on slow reward functions. This setup should scale well, both horizontally and vertically. Horizontally: We can have more actors, and each will manage their own rewards. Vertically: because we call the reward function asynchronously, we can have slower reward functions with minimal impact to overall throughput. 
2. This is conceptually cleaner, as now, when we have a completion done, it has all the information needed to process it. 

Runs: 

- Single GPU GRPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBDKS1D16Q9NT88RXJ69E4J0?taskId=01KBDKS1D8FFTDNRGFA7X00Z62&jobId=01KBDKS1HEXJKQ203Q19HMF9DN)
- Multi-node GRPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBDR25VNRPWXG0E114E9KB4F?taskId=01KBDR25VQ666ZAFQY4C03RR5E&jobId=01KBDR25Z8RXF7BFJ2JJ5ZDZ2S)
- Single GPU GRPO with tools: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBDKGVX0AJ4W5JDACDTR91MH?taskId=01KBDKGVX5G8A8FYYXV0M1FSA8&jobId=01KBDKGW0VYH19T0ANKQPBMX9W)

---

> [!NOTE]
> Moves reward calculation into vLLM actors using an async RewardConfig-built function, attaching scores/metrics to GenerationResult and updating the training/eval pipeline to consume them.
> 
> - **Runtime/Actors**:
>   - Compute rewards inside `LLMRayActor` after generation; build async `reward_fn` from new `RewardConfig` and attach `reward_scores`/`reward_metrics` to `GenerationResult`.
>   - `create_vllm_engines`/`LLMRayActor` now accept `reward_config`, `train_dataset`, `eval_dataset` to enable in-actor reward computation.
> - **Training pipeline**:
>   - Remove `reward_fn` plumbing from `grpo_fast`; `accumulate_inference_batches` and eval now read `result.reward_scores`/`reward_metrics` and derive stats from them.
>   - `create_model_and_optimizer` passes `RewardConfig` and datasets to engines.
> - **Ground truth utilities**:
>   - Add `apply_verifiable_reward` and `RewardConfig.build()` producing the reward function; extend metrics (per-verifier averages and correct rates).
> - **Data structures**:
>   - Extend `GenerationResult` with `reward_scores` and `reward_metrics`.
> - **Tests**:
>   - Update tests to construct `GenerationResult` with `reward_scores` and remove explicit `reward_fn` usage where applicable.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit eff4c591ad6903b7d56d91cd4e072bbaaae6d6ed. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/vllm_utils.py [hunk 1]
```diff
@@ -502,6 +505,68 @@ def add_request(actor: "LLMRayActor", request: PromptRequest) -> None:
...
+    result.reward_scores, result.reward_metrics = await compute_rewards(actor, result, dataset, is_eval)
+    results_queue = actor.eval_results_queue if is_eval else actor.results_queue
+    results_queue.put(result)
+
+
+async def compute_rewards(
+    actor: "LLMRayActor", result: GenerationResult, dataset: datasets.Dataset, is_eval: bool
+) -> tuple[list[float], dict]:
+    example = dataset[result.dataset_index]
+    decoded_responses = actor.llm_engine.tokenizer.batch_decode(result.responses)
```
[the remark this answers is not in the document]
Comment (hamishivi) [reply open_instruct/vllm_utils.py]: +1 on this, we should be explicit about special token handling so the reader knows what the reward code sees.

**Comment by finbarrtimbers** [comment]
Fixed the `skip_special_tokens=True` issue!
